src/core/mount.c

```diff
@@ -1280,23 +1280,22 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
         log_unit_full(u, f == MOUNT_SUCCESS ? LOG_DEBUG : LOG_NOTICE, 0,
                       "Mount process exited, code=%s status=%i", sigchld_code_to_string(code), status);
 
-        /* Note that mount(8) returning and the kernel sending us a mount table change event might happen
-         * out-of-order. If an operation succeed we assume the kernel will follow soon too and already change into the
-         * resulting state.  If it fails we check if the kernel still knows about the mount. and change state
-         * accordingly. */
+        /* Note that due to the io event priority logic, we can be sure the new mountinfo is loaded
+         * before we process the SIGCHLD for the mount command. */
 
         switch (m->state) {
 
         case MOUNT_MOUNTING:
-        case MOUNT_MOUNTING_DONE:
+                /* No new mountinfo for this mount point.
+                 * Someone raced us and we lost (or /bin/mount lies to us and is broken). */
+                assert(!m->from_proc_self_mountinfo);
```

Comment:
hmpf, please don't add assert()s like this for unit states. We previously had a lot of them in place, but they ended up breaking upgrades quite often, as serialization/deserialization between two different systemd versions frequently broke the assumptions made in these asserts, resulting in PID 1 hitting an assert...

I think we should be defensive here: yes, something is weird, but we should handle it in a graceful way, to a point where that's possible...

Reply:
Ah, ouch.  It was more about documentation, i.e. to formalize the comment.  I think the comment is enough.

I think transition across this commit won't break the assert (neither upgrade nor downgrade).  So I'm not too worried about what weird thing will happen.  But as a blanket rule it makes sense to remove the assert.